Support permission checkers for WebSockets Next

michalvavrik opened the pull request:
- so far there was no documented support for the checker methods and passing method arguments didn't work
- motivation for the support is this user that tried to use the checker methods https://github.com/quarkusio/quarkus/issues/40312#issuecomment-2575815962

mkouba commented on docs/src/main/asciidoc/websockets-next-reference.adoc:
```diff
@@ -883,6 +886,83 @@ quarkus.http.auth.permission.http-upgrade.paths=/end
+=== Secure server endpoints with permission checkers
+
+WebSocket endpoints can be secured with the xref:security-authorize-web-endpoints-reference.adoc#permission-checker[permission checkers].
+We strongly recommend to revise the <<secure-http-upgrade>> section and secure the HTTP upgrade like in the example below:
```
What exactly does this sentence mean? Should a user always use `@PermissionsAllowed`?

michalvavrik replied:
I tried to say to users: if you don't need to secure payload, you should secure HTTP upgrade. I know you have example above that allows to secure `@OnTextMessage` with `@RolesAllowed` and handle failure, but I don't believe it is much practical, because unless you close the connection, you will have connection for endpoint that can't be used. Maybe I should rephrase this. Suggestions?

michalvavrik replied:
>  Should a user always use @PermissionsAllowed

any standard security annotation is fine

michalvavrik replied:
> I wonder if we should move the HTTP upgrade section before the @OnTextMessage with @RolesAllowed example?

yes, I'll do it and try to rewrite the sentence as well.

michalvavrik replied:
moved the sections and adjusted text, please have a look

sberyozkin commented on docs/src/main/asciidoc/websockets-next-reference.adoc:
```diff
@@ -185,7 +185,8 @@ public class MyWebSocket {
 ==== Request context
 
-If an endpoint is annotated with `@RequestScoped`, or with a security annotation (such as `@RolesAllowed`), or depends directly or indirectly on a `@RequestScoped` bean, or on a bean annotated with a security annotation, then each WebSocket endpoint callback method execution is associated with a new _request context_.
+If an endpoint is annotated with `@RequestScoped`, or with a security annotation (such as `@RolesAllowed`) that does not secure an HTTP upgrade,
```
```suggestion
If an endpoint is annotated with `@RequestScoped`, or a security annotation such as `@RolesAllowed` that does not secure an HTTP upgrade,
```

sberyozkin replied:
It may be worth splitting this sentence into something like: `Each WebSocket endpoint callback method execution is associated with a new _request context_ in the following cases:` and list them, as otherwise this sentence looks quite complex, Rolfe would definitely prefer to split it :-)

michalvavrik replied:
I went with the suggestion to create a list, but didn't exactly go word by word. Please have a look and let's discuss it in a new comment thread (linked to changed text, so that I know what to change), thanks

sberyozkin commented on docs/src/main/asciidoc/websockets-next-reference.adoc:
```diff
@@ -185,7 +185,8 @@ public class MyWebSocket {
 
-If an endpoint is annotated with `@RequestScoped`, or with a security annotation (such as `@RolesAllowed`), or depends directly or indirectly on a `@RequestScoped` bean, or on a bean annotated with a security annotation, then each WebSocket endpoint callback method execution is associated with a new _request context_.
+If an endpoint is annotated with `@RequestScoped`, or with a security annotation (such as `@RolesAllowed`) that does not secure an HTTP upgrade,
+or depends directly or indirectly on a `@RequestScoped` bean, or on a bean annotated with a security annotation, then each WebSocket endpoint callback method execution is associated with a new _request context_.
```
```suggestion
or depends on a `@RequestScoped` bean, or on a bean annotated with a security annotation, then each WebSocket endpoint callback method execution is associated with a new _request context_.
```

sberyozkin commented on docs/src/main/asciidoc/websockets-next-reference.adoc:
```diff
@@ -832,11 +833,13 @@ public class Endpoint {
 An HTTP upgrade is secured when standard security annotation is placed on an endpoint class or an HTTP Security policy is defined.
 The advantage of securing HTTP upgrade is less processing, the authorization is performed early and only once.
-You should always prefer HTTP upgrade security unless, like in th example above, you need to perform action on error.
+You should always prefer HTTP upgrade security unless, like in the example above, you need to perform action on error
```
```suggestion
You should always prefer HTTP upgrade security unless, like in the example above, you need to perform an action on error
```

sberyozkin commented on docs/src/main/asciidoc/websockets-next-reference.adoc:
```diff
@@ -832,11 +833,13 @@ public class Endpoint {
 The advantage of securing HTTP upgrade is less processing, the authorization is performed early and only once.
-You should always prefer HTTP upgrade security unless, like in th example above, you need to perform action on error.
+You should always prefer HTTP upgrade security unless, like in the example above, you need to perform action on error
+or you need to perform a security check based on the payload.
```
```suggestion
or a security check based on the payload.
```

sberyozkin commented on docs/src/main/asciidoc/websockets-next-reference.adoc:
```diff
@@ -883,6 +886,83 @@ quarkus.http.auth.permission.http-upgrade.paths=/end
+
+}
+----
+<1> A permission checker authorizing an HTTP upgrade must declare exactly one method parameter, the `SecurityIdentity`.
```
Is it too restrictive to require a single parameter, of specific type ? HTTP upgrade can contain query parameters, for example, for the custom ticketing system, https://github.com/quarkiverse/quarkus-langchain4j/tree/main/samples/secure-sql-chatbot#custom-websocket-ticket-system

michalvavrik replied:
This is only valid for HTTP upgrade.

I don't mind to support more arguments, but this restriction comes from the fact that it is a `@PermissionsAllowed` placed on a class, so only `SecurityIdentity` is supported in Quarkus Security. The reason for this is that in general, where would you take these other arguments from if it is not applied on a method? From CDI or .... We can implement that OR part. Which do you have in mind? If it is about `RoutingContext`, I presume it will be available on `SecurityIdentity` if it is OIDC or basic for auth.

michalvavrik replied:
maybe showing how to access routing context in the example would be useful

michalvavrik replied:
I was also thinking that if user needs something really special, they could secure `@OnOpen` (cc @mkouba ) which would be very similar, just bit more processing but allows you to inject connection. But I was too tired yesterday to create test so I let it be.

sberyozkin replied:
Hi @michalvavrik Sure, custom ticketing system is not a typical case, let's keep it as is

michalvavrik replied:
@sberyozkin I was wrong, not sure why I remember it incorrectly (maybe I am not looking right), but I only found OIDC identity provider adds the `RoutingContext`. So I made couple of changes to ensure that `RoutingContext` is always propagated from authentication request (where we know it is as builtin auth mechanisms add it) to the identity on which permission check is performed.

I am not exactly sure how you will feel about changes, I think they are not super pretty but acceptable. if you have strong opinion, let me know and I'll revert it.

michalvavrik commented:
I pushed it a little sooner than I meant, I'll make couple more changes in a gist.

michalvavrik commented:
hey @mkouba @sberyozkin , thanks for the review, I addressed your comments, but I didn't just apply your suggestions so there is more changes (improvements hopefully) to docs. I also improved test coverage. Please have a look and let's start the review again.

michalvavrik commented:
> A few Elytron security tests are failing

I think their assertions are not written well, I'll fix the tests.

michalvavrik commented:
Done, there is a permission checker in `RootResource` so all the identities had extra `RoutingContext` (which is IMHO fine), but the test didn't expect that.

sberyozkin commented on docs/src/main/asciidoc/websockets-next-reference.adoc:
```diff
@@ -185,8 +185,12 @@ public class MyWebSocket {
 
-If an endpoint is annotated with `@RequestScoped`, or with a security annotation (such as `@RolesAllowed`), or depends directly or indirectly on a `@RequestScoped` bean, or on a bean annotated with a security annotation, then each WebSocket endpoint callback method execution is associated with a new _request context_.
-The request context is active during endpoint callback invocation.
+Each WebSocket endpoint callback method execution is associated with a new _request context_ (activate CDI request context) if the endpoint is:
```
```suggestion
Each WebSocket endpoint callback method execution is associated with a new CDI _request context_ if the endpoint is:
```

I'm just not sure how to change `activate CDI request context` since `activate` is a verb

sberyozkin commented on docs/src/main/asciidoc/websockets-next-reference.adoc:
```diff
@@ -783,6 +787,67 @@ class MyBean {
 === Security
 
+Security capabilities are provided by the Quarkus Security extension.
+Any xref:security-identity-providers.adoc[Identity provider] can be used to convert authentication credentials on initial HTTP request into a `SecurityIdentity` instance.
```
```suggestion
Any xref:security-identity-providers.adoc[Identity provider] can be used to convert authentication credentials on the initial HTTP request into a `SecurityIdentity` instance.
```

sberyozkin commented on docs/src/main/asciidoc/websockets-next-reference.adoc:
```diff
@@ -783,6 +787,67 @@ class MyBean {
+The `SecurityIdentity` is then associated with the websocket connection.
+Authorization options are demonstrated in following sections.
+
+NOTE: When OpenID Connect extension is used and token expires, Quarkus automatically closes connection.
```
```suggestion
NOTE: When OpenID Connect extension, `quarkus-oidc`, is used and token expires, Quarkus automatically closes connection.
```

Rolfe just did a similar update in one of his PRs

sberyozkin commented on docs/src/main/asciidoc/websockets-next-reference.adoc:
```diff
@@ -783,6 +787,67 @@ class MyBean {
+[[secure-http-upgrade]]
+==== Secure HTTP upgrade
+
+An HTTP upgrade is secured when standard security annotation is placed on an endpoint class or an HTTP Security policy is defined.
```
```suggestion
An HTTP upgrade is secured when a standard security annotation is placed on an endpoint class or an HTTP Security policy is defined.
```

sberyozkin commented on docs/src/main/asciidoc/websockets-next-reference.adoc:
```diff
@@ -828,60 +893,108 @@ public class Endpoint {
 
-NOTE: When OpenID Connect extension is used and token expires, Quarkus automatically closes connection.
+WebSocket endpoints can be secured with the xref:security-authorize-web-endpoints-reference.adoc#permission-checker[permission checkers].
+We recommend to <<secure-http-upgrade>> rather than individual endpoint methods. For example:
```
I'm not sure about the recommendation. What if one requires an authentication at the upgrade time, and different permissions for different endpoint methods ?

michalvavrik replied:
For now, there is exactly one endpoint for one connection.

michalvavrik replied:
As for endpoint methods, well, if you can open connection I think you can also get messages etc.

sberyozkin replied:
Sorry, looks like I'm missing something. Does it work now:

```
Authenticated <1>
@WebSocket(path = "/end")
public class Endpoint {
    @Inject
    SecurityIdentity currentIdentity;
    @OnOpen
    String open() {
        return "ready";
    }
    @RolesAllowed("admin")
    @OnTextMessage
    String echo(String message) {
        return message;
    }
}
```
?

michalvavrik replied:
It does work, but every time `echo` is invoked, allowed roles are checked. The `SecurityIdentity` roles are immutable, therefore I'd check them once instead of many times.

michalvavrik replied:
I'll put it another way @sberyozkin , because maybe it is me who is missing something. What do you get by annotating `echo` with `@RolesAllowed` instead of `Endpoint`. do you have in mind a scenario that would profit from that?

sberyozkin commented on docs/src/main/asciidoc/websockets-next-reference.adoc:
```diff
@@ -828,60 +893,108 @@ public class Endpoint {
+----
+<1> A permission checker authorizing an HTTP upgrade must declare exactly one method parameter, the `SecurityIdentity`.
+
+If the security check performs payload-aware authorization, it is also possible to run check on every message. For example:
```
It is also possible to run security checks on every message. For example, a message payload can be accessed like this:

sberyozkin replied:
```suggestion
It is also possible to run security checks on every message. For example, a message payload can be accessed like this:
```

sberyozkin commented on extensions/security/runtime/src/main/java/io/quarkus/security/runtime/QuarkusSecurityIdentity.java:
```diff
@@ -149,6 +162,7 @@ public static class Builder {
         Set<Permission> permissions = new HashSet<>();
         Map<String, Object> attributes = new HashMap<>();
         List<Function<Permission, Uni<Boolean>>> permissionCheckers = new ArrayList<>();
+        List<BiFunction<SecurityIdentity, Permission, Uni<Boolean>>> permissionCheckersWithIdentity = new ArrayList<>();
```
I'm not sure putting it here is great, this builder is really for the users who write the augmentation, IMHO the original code where the permission checker is added after the identity was created was fine

michalvavrik replied:
I usually don't use context suggestions for methods that are not available (like private methods, or package-private when outside of the package) in my IDE. That is why I don't think it can confuse users. I only use Intellij Idea though, maybe eclipse is not that smart?

Sorry if it is too obvious, but let me put into context why I have done it - it allows you to propagate identity attributes without building identity twice inside one augmentor.

in any case, I agree it is not super nice, I can drop it if you deem it necessary.

michalvavrik replied:
also, I'd much prefer to make RoutingContext injectable as method parameter. The reason why I didn't do it is that it is very WS Next http upgrade specific. in any other case I am aware of you already have `RoutingContext` injectable from CDI.

michalvavrik replied:
I'll create  a new class and avoid adding a new method to `QuarkusSecurityIdentity` builder in order to unblock this PR. I'll push it in an hour.

sberyozkin replied:
Sorry Michal, but indeed, it will most likely look nicer, IMHO the builder should only offer code/methods that can be of of help when users work with this builder in their custom augmentors

michalvavrik replied:
np, done

sberyozkin commented on docs/src/main/asciidoc/websockets-next-reference.adoc:
```diff
@@ -783,6 +787,67 @@ class MyBean {
+<1> Initial HTTP handshake ends with the 401 status for anonymous users.
+You can also redirect the handshake request on authorization failure with the `quarkus.websockets-next.server.security.auth-failure-redirect-url` configuration property.
+
+IMPORTANT: HTTP upgrade is only secured when a security annotation is declared on an endpoint class next to the `@WebSocket` annotation.
```
This is probably a possible follow up discussion, but it appears unusual that the endpoint methods require method specific security annotations, IMHO, it would make sense to inherit the class level annotation since the security identity is bound to the connection anyway. REST endpoint users can only have a class-level security annotation and have all endpoint methods secured...

michalvavrik replied:
I know it works differently (and the difference in behavior is documented), but so far I miss scenario where it would be useful. If HTTP Upgrade is permitted by class-level annotations, nothing will change because identity will not change .

Except if you need to check for example `@OnTextMessage` method arguments, that was my motivation for this PR. I believe there will be users that want to only permit access if payload is OK for given identity (and there will also be different users who will do it on a service level instead... :-) _.

I am open to discussion, but this is not a new behavior, I just moved docs to further up. Can you open a separate discussion (GH issue or discussion) if you think there is more into this, please?

sberyozkin commented on docs/src/main/asciidoc/websockets-next-reference.adoc:
```diff
@@ -783,6 +787,67 @@ class MyBean {
+==== Secure HTTP upgrade
+
+An HTTP upgrade is secured when a standard security annotation is placed on an endpoint class or an HTTP Security policy is defined.
+The advantage of securing HTTP upgrade is less processing, the authorization is performed early and only once.
```
IMHO lines 801 and 802 may need to be reworked. IMHO it is really a question whether users want to have a secure upgrade or not as opposed to optimizing the security checks (or securing callback methods only), since the HTTP upgrade is the main entry point... 

May be we can just drop lines 801, 802 ? Line 800 is good

michalvavrik replied:
I don't mind dropping 801, but I think it is important for users to understand the advantage of securing HTTP upgrade. The advantage is that you perform security check once instead of on every message, which in case of WS can be  a lot. Maybe you can suggest rephrasing 802, please?

sberyozkin replied:
@michalvavrik Sure, let us just move on with this PR and consider tuning the text which was already there before later

sberyozkin commented:
Thanks @michalvavrik, not sure about that :-), but thanks anyway :-)

sberyozkin reviewed (approved):
Thanks @michalvavrik, let's wait for @mkouba to approve as well

michalvavrik commented:
That `TestSecurityTestCase` failed again after the latest changes, I'll look into it.

michalvavrik commented:
Sorry, stupid mistake, fixed now.

mkouba commented on docs/src/main/asciidoc/websockets-next-reference.adoc:
```diff
@@ -783,6 +787,67 @@ class MyBean {
+
+An HTTP upgrade is secured when a standard security annotation is placed on an endpoint class or an HTTP Security policy is defined.
+The advantage of securing HTTP upgrade is less processing, the authorization is performed early and only once.
+You should always prefer HTTP upgrade security unless, like in the example above, you need to perform an action on error or a security check based on the payload.
```
> like in the example abov

Which example is it?

michalvavrik replied:
now it is example in the next section, I'll refer to that section in gist, good catch

michalvavrik committed:
Support permission checkers for WebSockets Next

mkouba reviewed (approved):
It looks good in general and I believe that @sberyozkin carefully reviewed the security bits ;-).

michalvavrik commented:
MicroProfile OpenTelemetry TCK failure seems unrelated. IMHO this is ready for merge.
